core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -133,51 +131,155 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
-                    NO_DELETES,
-                    context.schemaAsString(),
-                    context.specAsString(),
-                    context.residuals());
-
-              default:
-                throw new IllegalArgumentException("Unexpected entry status: " + entry.status());
-            }
-          });
+      CloseableIterable<ChangelogScanTask> tasks =
+          CloseableIterable.transform(
+              entries,
+              entry -> {
+                long entrySnapshotId = entry.snapshotId();
+                DataFile dataFile = entry.file().copy(context.shouldKeepStats());
+                DeleteFile[] addedDeleteFiles = filterAdded(context.deletes().forEntry(entry));
+
+                switch (entry.status()) {
+                  case ADDED:
+                    if (entrySnapshotId == snapshotId) {
+                      return new BaseAddedRowsScanTask(
+                          changeOrdinal,
+                          snapshotId,
+                          dataFile,
+                          addedDeleteFiles,
+                          context.schemaAsString(),
+                          context.specAsString(),
+                          context.residuals());
+                    } else {
+                      // the data file is added before the snapshot we're processing
+                      if (addedDeleteFiles.length == 0) {
+                        return new DummyChangelogScanTask(changeOrdinal, snapshotId);
+                      } else {
+                        return new BaseDeletedRowsScanTask(
+                            changeOrdinal,
+                            snapshotId,
+                            dataFile,
+                            addedDeleteFiles,
+                            // not used
+                            filterExisting(context.deletes().forEntry(entry)),
+                            context.schemaAsString(),
+                            context.specAsString(),
+                            context.residuals());
+                      }
+                    }
+
+                  case DELETED:
+                    if (entrySnapshotId == snapshotId) {
+                      return new BaseDeletedDataFileScanTask(
+                          changeOrdinal,
+                          snapshotId,
+                          dataFile,
+                          filterExisting(context.deletes().forEntry(entry)),
+                          context.schemaAsString(),
+                          context.specAsString(),
+                          context.residuals());
+                    } else {
+                      return new DummyChangelogScanTask(changeOrdinal, snapshotId);
+                    }
+
+                  case EXISTING:
+                    if (addedDeleteFiles.length == 0) {
+                      return new DummyChangelogScanTask(changeOrdinal, snapshotId);
+                    } else {
+                      return new BaseDeletedRowsScanTask(
```

Comment:
Would this cause a full table scan for every snapshot in the snapshot range we are checking when there is an equality delete file present?

Reply:
Note that a BaseDeletedRowsScanTask is only created if there are delete files *added in that snapshot* that may apply to the data file.
Supposing that there is an equality delete file added in that snapshot, whether a full table scan happens for that snapshot depends on the equality deletes and how well `DeleteFileIndex.canContainEqDeletesForFile(DataFile, EqualityDeleteFile)` is able to filter them. If the equality deletes are for a partition, or even if global but can be excluded using ranges, then a full table scan may not happen. But yes, depending on the equality deletes it could happen.
Unfortunately, I don't see an alternative.